lib/exercism/github_auth.rb

```diff
@@ -1,40 +1,49 @@
-class Github
-  def self.login_url(q)
+class GitHubAuth
+  class UnauthorizedUser < StandardError; end
+  USER_AGENT = 'github.com:exercism/exercism.io'
+
+  def self.login_url(q = {})
     "https://github.com/login/oauth/authorize?#{q.map { |k, v| '%s=%s' % [k, v] }.join('&')}"
   end
 
-  def self.connect_to(url)
+    def self.connect_to(url)
     Faraday.new(url: url) do |c|
       c.use Faraday::Response::Logger
       c.use Faraday::Adapter::NetHttp
     end
   end
 
-  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
-  def self.authenticate(code, client_id, client_secret)
-    conn = connect_to('https://github.com')
-
-    response = conn.post do |req|
-      req.url '/login/oauth/access_token'
-      req.headers['Content-Type'] = 'application/json'
-      req.headers['Accept'] = 'application/json'
-      req.headers['User-Agent'] = user_agent
-      req.body = { client_id: client_id, client_secret: client_secret, code: code }.to_json
-    end
-    access_token = JSON.parse(response.body)['access_token']
-
+  def self.user_info(access_token)
     conn = connect_to 'https://api.github.com'
     response = conn.get do |req|
       req.url '/user'
-      req.headers['User-Agent'] = user_agent
+      req.headers['User-Agent'] = USER_AGENT
       req.params['access_token'] = access_token
     end
+    raise UnauthorizedUser unless response.success?
     result = JSON.parse(response.body)
-    [result['id'], result['login'], result['email'], result['avatar_url']]
+    GitHubUser.new(result)
   end
-  # rubocop:enable Metrics/AbcSize, Metrics/MethodLength
 
-  def self.user_agent
-    'github.com:exercism/exercism.io'
+  def initialize(code, client_id, client_secret)
+    @code = code
+    @client_id = client_id
+    @client_secret = client_secret
   end
+
+  def access_token
+    conn = self.class.connect_to('https://github.com')
+
+    response = conn.post do |req|
+      req.url '/login/oauth/access_token'
+      req.headers['Content-Type'] = 'application/json'
+      req.headers['Accept'] = 'application/json'
+      req.headers['User-Agent'] = USER_AGENT
+      req.body = { client_id: client_id, client_secret: client_secret, code: code }.to_json
+    end
+    JSON.parse(response.body)['access_token']
+  end
```

Comment:
What should this method do? It looks like it's doing at least 3 things.
Perhaps splitting the responsibilities into their own methods might  make things clearer.